fontbe/src/colr.rs

```diff
@@ -240,6 +240,85 @@ fn to_colr_paint(
     }
 }
 
+fn add_or_extend_clip(clips: &mut Vec<Clip>, quantization: i16, gid: GlyphId16, glyph: &Glyph) {
+    let bbox = glyph.data.bbox().unwrap_or_default();
+    // Quantize the bbox to maximize clipbox reuse
+    let next_clip = quantize_bbox(&bbox, quantization);
+    let next_clip = ClipBox::format_1(
+        next_clip.x_min.into(),
+        next_clip.y_min.into(),
+        next_clip.x_max.into(),
+        next_clip.y_max.into(),
+    );
+    if let Some(curr) = clips.last_mut()
+        && curr.end_glyph_id.to_u32() + 1 == gid.to_u32()
+        && *curr.clip_box == next_clip
+    {
+        // wow wow wow, a run!
+        curr.end_glyph_id = gid;
+        return; // DONE
+    }
+
+    // Evidently we didn't make a run
+    clips.push(Clip::new(gid, gid, next_clip));
+}
+
+fn is_paint_glyph_solid(paint: &ir::Paint) -> bool {
+    if let ir::Paint::Glyph(paint_glyph) = paint
+        && let ir::Paint::Solid(_) = &paint_glyph.paint
+    {
+        true
+    } else {
+        false
+    }
+}
+
+fn colr_v0_compatible(paint: &ir::Paint) -> bool {
+    if is_paint_glyph_solid(paint) {
+        return true;
+    }
+    if let ir::Paint::Layers(layers) = paint
+        && layers.iter().all(is_paint_glyph_solid)
+    {
+        return true;
+    }
+    false
+}
+
+fn new_colr0(
+    palette: &ColorPalettes,
+    glyph_order: &GlyphOrder,
+    glyph_name: &GlyphName,
+    paint: &[&ir::PaintGlyph],
+    layers: &mut Vec<Layer>,
+) -> Result<BaseGlyph, Error> {
+    let gid = glyph_order.glyph_id(glyph_name).expect("Prevalidated");
+    let candidate_layers = paint
+        .iter()
+        .map(|p| {
+            let ir::Paint::Solid(solid) = &p.paint else {
+                panic!("Prevalidated");
```

Comment:
for code that should be genuinely unreachable `unreachable!()`, which is equivalent to `panic!` but better communicates your intention. (there's also the unsafe `unreachable_unchecked`, for when you're actually telling the compiler that a given branch can be removed.

If we _do_ ever think it's possible to hit this we might want to include the gid or glyph name in the message?

Reply:
So updated